add externalTrafficPolicy toggle

furstblumier opened the pull request:
See #144

cfis commented on charts/docker-mailserver/values.yaml:
```diff
@@ -298,6 +298,11 @@ service:
+  ## - Local
+  ## - Cluster
+  ## Set it to "Local" when used with type "LoadBalancer", unless you have a good reason not to.
+  # externalTrafficPolicy: Local
```
Maybe uncomment this line out? And then in service.yaml have:

```
externalTrafficPolicy: {{ .Values.service.externalTrafficPolicy }}
```

furstblumier replied:
Alrighty!

kaktus42 commented:
Great effort!

The way you do it now, it changes the original behaviour non-transparently.

When it's used with `.Values.service.type: NodePort`, it will not set `externalTrafficPolicy` to `Cluster` anymore.

Would be better to use an override and check if the override is given, else use original logic.

furstblumier commented:
> Great effort!
> 
> The way you do it now, it changes the original behaviour non-transparently.
> 
> When it's used with `.Values.service.type: NodePort`, it will not set `externalTrafficPolicy` to `Cluster` anymore.
> 
> Would be better to use an override and check if the override is given, else use original logic.

I updated the comment in the values to reflect it a bit better, but essentially I removed the logic on purpose to also get edge cases, where the logic fails, to work. Maybe someone has a very good reason to do NodePort + Local or some other combination.

cfis commented:
@kaktus42 - Not seeing a change? Service type has a default of "LoadBalancer" which  would set "externalTrafficPolicy " to local, which is now its default.

Having said that, should the service default should be "ClusterIP"?

@furstblumier can you update the Chart version - that will make the lint command work.

furstblumier commented:
@cfis The issue isn't the chart version, but I think I can fix it. I'll also set the default to "ClusterIP" I guess.

furstblumier commented:
Sorry, I read it wrong. Should be good now

cfis commented:
Lint test failed....

furstblumier commented:
Bumped the chart version and hopefully fixed ci this time

cfis commented:
Same error as before:

> namespace/docker-mailserver-oi25urtesg created
> Error: INSTALLATION FAILED: 1 error occurred:
> 	* Service "docker-mailserver-oi25urtesg" is invalid: spec.externalTrafficPolicy: Invalid value: "Local": may only be set when `type` is 'NodePort' or 'LoadBalancer'

furstblumier commented:
Alright, now it should be fine. I wasn't aware that with service type "ClusterIP" you aren't "allowed" to set externalTrafficPolicy at all. I adjusted the values accordingly.

I also implemented the old default logic again. If you don't set "externalTrafficPolicy" it will use the previous defaults.

furstblumier commented on charts/docker-mailserver/templates/service.yaml:
```diff
@@ -26,11 +26,17 @@ metadata:
   ## If a load balancer is being used, ensure that the newer type of LB that passes along IP information is used
   ## rather than the legacy one.
-  {{- if eq .Values.service.type "LoadBalancer" }}
+  {{- if or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer") }}
```
Only do this if we are using LB or NodePort

furstblumier commented on charts/docker-mailserver/templates/service.yaml:
```diff
@@ -26,11 +26,17 @@ metadata:
   ## rather than the legacy one.
-  {{- if eq .Values.service.type "LoadBalancer" }}
+  {{- if or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer") }}
+  {{- if .Values.service.externalTrafficPolicy }}
```
if we have an explicit value set, use that

furstblumier commented on charts/docker-mailserver/templates/service.yaml:
```diff
@@ -26,11 +26,17 @@ metadata:
+  {{- if .Values.service.externalTrafficPolicy }}
+  externalTrafficPolicy: {{ .Values.service.externalTrafficPolicy }}
+  {{- end }}
+  {{- if and (not .Values.service.externalTrafficPolicy) (eq .Values.service.type "LoadBalancer") }}
```
if we don't have an explicit value set AND the type is Loadbalancer, use Local

furstblumier commented on charts/docker-mailserver/templates/service.yaml:
```diff
@@ -26,11 +26,17 @@ metadata:
   externalTrafficPolicy: Local
-  {{- else if eq .Values.service.type "NodePort" }}
+  {{- end }}
+  {{- if and (not .Values.service.externalTrafficPolicy) (eq .Values.service.type "NodePort") }}
```
if we don't have an explicit value set AND the type is NodePort, use Cluster

furstblumier commented:
@cfis Would it be possible for you to review this? I'd much obliged, since I'd love to get rid of my workaround in my homelab...

cfis commented:
Thanks for the reminder, I had forgotten that I didn't merge this yet.
